java/arcs/core/storage/StorageProxy.kt

```diff
@@ -58,71 +65,109 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
+            CoroutineScope(coroutineContext).launch { action() }
+        }
+    }
+
+    /**
+     * Add a [Handle] `onUpdate` action, associated with a [Handle] name.
+     *
+     * If the [StorageProxy] is in its initial state, the first call to any of the action methods
+     * will trigger a request for synchronization.
+     */
     suspend fun addOnUpdate(handleName: String, action: suspend (value: T) -> Unit) {
-        callbackMutex.withLock {
+        addAction {
             onUpdateActions.getOrPut(handleName) { mutableListOf() }.add(action)
         }
     }
 
     /**
-     * Add a [Handle] `onSync` action, associated with a [Handle] name.
+     * Add a [Handle] `onDesync` action, associated with a [Handle] name.
      *
-     * If the [StorageProxy] is synchronized when the action is added, it will be called
+     * If the [StorageProxy] is desynchronized when the action is added, it will be called
      * immediately.
      *
-     * If not, the [StorageProxy] will request synchronization.
+     * If the [StorageProxy] is in its initial state, the first call to any of the action methods
+     * will trigger a request for synchronization.
      */
-    suspend fun addOnSync(handleName: String, action: () -> Unit) {
-        val isAlreadySynchronized = syncMutex.withLock {
-            callbackMutex.withLock {
-                onSyncActions.getOrPut(handleName) { mutableListOf() }.add(action)
-            }
-            isSynchronized
+    suspend fun addOnDesync(handleName: String, action: () -> Unit) {
+        val currentState = addAction {
+            onDesyncActions.getOrPut(handleName) { mutableListOf() }.add(action)
         }
-        if (isAlreadySynchronized) {
+        if (currentState == ProxyState.DESYNC) {
             CoroutineScope(coroutineContext).launch { action() }
-        } else {
-            requestSynchronization()
         }
     }
 
     /**
-     * Add a [Handle] `onDesync` action, associated with a [Handle] name.
+     * Add a [Handle] `onResync` action, associated with a [Handle] name.
      *
-     * If the handle is not synchronized when the action is added, it will be called immediately.
+     * If the [StorageProxy] is in its initial state, the first call to any of the action methods
+     * will trigger a request for synchronization.
      */
-    suspend fun addOnDesync(handleName: String, action: () -> Unit) {
-        val runNow = syncMutex.withLock {
+    suspend fun addOnResync(handleName: String, action: () -> Unit) {
+        addAction {
+            onResyncActions.getOrPut(handleName) { mutableListOf() }.add(action)
+        }
+    }
+
+    /**
+     * Run the `add` function to add a storage event action method.
+     *
+     * If the [StorageProxy] is in the INIT state, send a sync request and move to AWAITING_SYNC.
+     */
+    private suspend fun addAction(add: () -> Unit): ProxyState {
+        var needsSync = false
+        val currentState = syncMutex.withLock {
```

Comment:
return a pair and unpack with `val (currentState, needsSync)` here?

Reply:
Not sure that's much of an improvement - either you have a local var inside the mutex block anyway, or you have: 
```
if (state == ProxyState.INIT) {
    state = ProxyState.AWAITING_SYNC
    Pair(state, true)
} else {
    Pair(state, false)
}
```

..which to me seems less clear.

Reply:
I don't feel that strongly about it, but personally I prefer the pattern that requires defining a mutable variable outside of the block and mutating inside.

Reply:
.. isn't that the opposite of your original suggestion?

Reply:
Oh, yes... I typed the exact opposite of what was in my head. I guess I meant to say "that does not require". Or maybe my subconscious prefers what you have.